Objects/exceptions.c

```diff
@@ -629,6 +629,450 @@ ComplexExtendsException(PyExc_BaseException, SystemExit, SystemExit,
+    Py_ssize_t num_excs = PySequence_Size(excs);
+    if (num_excs < 0) {
+        return NULL;
+    }
+    else if (num_excs == 0) {
+        return Py_NewRef(Py_None);
+    }
+
+    PyObject *eg = PyObject_CallMethod(
+        (PyObject*)orig, "derive", "(O)", excs);
+    if (!eg) {
+        return NULL;
+    }
+
+    if (!PyObject_TypeCheck(eg, (PyTypeObject *)PyExc_BaseExceptionGroup)) {
+        PyErr_SetString(PyExc_TypeError,
+            "derive must return an instance of BaseExceptionGroup");
+        goto error;
+    }
+
+    PyObject *tb = PyException_GetTraceback((PyObject*)orig);
+    if (tb) {
+        int res = PyException_SetTraceback(eg, tb);
+        Py_XDECREF(tb);
+        if (res == -1) {
+            goto error;
+        }
+    }
+    PyObject *context = PyException_GetContext((PyObject*)orig);
+    if (context) {
+        PyException_SetContext(eg, context);
+    }
+    PyObject *cause = PyException_GetCause((PyObject*)orig);
+    if (cause) {
+        PyException_SetCause(eg, cause);
+    }
+    return eg;
+error:
+    Py_XDECREF(eg);
+    return NULL;
+}
+
+enum _exceptiongroup_split_matcher_type {
+    /* Exception type or tuple of thereof */
+    EXCEPTION_GROUP_MATCH_BY_TYPE = 0,
+    /* A PyFunction returning True for matching exceptions */
+    EXCEPTION_GROUP_MATCH_BY_PREDICATE = 1,
+    /* An instance or container thereof, checked with equality */
+    EXCEPTION_GROUP_MATCH_INSTANCES = 2
+};
+
+struct _exceptiongroup_matcher {
+    enum _exceptiongroup_split_matcher_type type;
+    PyObject *value;
+};
+
+static int _set_matcher_type(
+    struct _exceptiongroup_matcher *matcher)
+{
+    /* the python API supports only BY_TYPE and BY_PREDICATE */
+    if (PyExceptionClass_Check(matcher->value) ||
+        PyTuple_CheckExact(matcher->value)) {
+        matcher->type = EXCEPTION_GROUP_MATCH_BY_TYPE;
+        return 0;
+    }
+    else if (PyFunction_Check(matcher->value)) {
+        matcher->type = EXCEPTION_GROUP_MATCH_BY_PREDICATE;
+        return 0;
+    }
+    PyErr_SetString(
+        PyExc_TypeError,
+        "expected a function, exception type or tuple of exception types");
+    return -1;
+}
+
+static int exceptiongroup_split_check_match(
+    PyObject *exc, const struct _exceptiongroup_matcher *matcher)
+{
+    switch (matcher->type) {
+    case EXCEPTION_GROUP_MATCH_BY_TYPE: {
+        return PyErr_GivenExceptionMatches(exc, matcher->value);
+    }
+    case EXCEPTION_GROUP_MATCH_BY_PREDICATE: {
+        PyObject *exc_matches = PyObject_CallOneArg(matcher->value, exc);
+        if (exc_matches == NULL) {
+            return -1;
+        }
+        int is_true = PyObject_IsTrue(exc_matches);
+        Py_DECREF(exc_matches);
+        return is_true;
+    }
+    case EXCEPTION_GROUP_MATCH_INSTANCES: {
+        if (PySequence_Check(matcher->value)) {
+            return PySequence_Contains(matcher->value, exc);
+        }
+        else {
+            return matcher->value == exc;
+        }
+    }
+    }
```

Comment:
Please choose to ignore this one, but IMO this is more readable

```suggestion
    case EXCEPTION_GROUP_MATCH_INSTANCES:
        if (PySequence_Check(matcher->value)) {
            return PySequence_Contains(matcher->value, exc);
        }
        return matcher->value == exc;
     }
```